File: stacks&queues/QueueArray.java
package stacks;

public class QueueArray {
    private int[] arr;     
    private int front;     
    private int rear;      
    private int size;      
    private int capacity;  

    // constructor
    public QueueArray(int capacity) {
        this.capacity = capacity;
        arr = new int[capacity];
        front = 0;
        rear = -1;
        size = 0;
    }

    // add element to queue
    public void enqueue(int value) {
        if (size == capacity) {
            System.out.println("Queue overflow!");
            return;
        }
        rear = (rear + 1) % capacity; // circular increment
        arr[rear] = value;
        size++;
    }

    // remove element from queue
    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue underflow!");
            return -1;
        }
        int value = arr[front];
        front = (front + 1) % capacity; // circular increment
        size--;
        return value;
    }

    // check if queue is empty
    public boolean isEmpty() {
        return size == 0;
    }

    // peek at front element
    public int peek() {
        if (isEmpty()) {
            System.out.println("Queue is empty!");
            return -1;
        }
        return arr[front];
    }

    // return size of queue
    public int size() {
        return size;
    }

    // print queue contents
    public void printQueue() {
        if (isEmpty()) {
            System.out.println("Queue is empty!");
            return;
        }
        int count = size;
        int i = front;
        while (count > 0) {
            System.out.print(arr[i] + " -> ");
            i = (i + 1) % capacity;
            count--;
        }
        System.out.println("null");
    }

    // test
    public static void main(String[] args) {
        QueueArray q1 = new QueueArray(5);

        q1.enqueue(7);
        q1.enqueue(2);
        q1.enqueue(0);
        q1.enqueue(4);
        q1.enqueue(6);

        q1.printQueue();
        System.out.println(q1.size());
        System.out.println(q1.dequeue());
        System.out.println(q1.isEmpty());
        q1.printQueue();
        System.out.println(q1.size());
        System.out.println(q1.peek());
    }
}

File: stacks&queues/queueArray.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct Queue {
    int *arr;       // array to store queue elements
    int capacity;   // maximum size of queue
    int front;      // index of front element
    int rear;       // index of rear element
    int size;       // current number of elements
} Queue;

// create a queue with given capacity
Queue* createQueue(int capacity) {
    Queue* newQueue = (Queue*)malloc(sizeof(Queue));
    if (!newQueue) {
        printf("Memory allocation failed!\n");
        return NULL;
    }
    newQueue->capacity = capacity;
    newQueue->arr = (int*)malloc(capacity * sizeof(int));
    if (!newQueue->arr) {
        printf("Memory allocation failed!\n");
        free(newQueue);
        return NULL;
    }
    newQueue->front = 0;
    newQueue->rear = -1;
    newQueue->size = 0;
    return newQueue;
}

// add element to queue
void enqueue(Queue* q, int value) {
    if (q->size == q->capacity) {
        printf("Queue overflow!\n");
        return;
    }
    q->rear = (q->rear + 1) % q->capacity; // circular increment
    q->arr[q->rear] = value;
    q->size++;
}

// remove element from queue
int dequeue(Queue* q) {
    if (q->size == 0) {
        printf("Queue underflow!\n");
        return -1; // sentinel value
    }
    int value = q->arr[q->front];
    q->front = (q->front + 1) % q->capacity; // circular increment
    q->size--;
    return value;
}

// check if queue is empty
bool isEmpty(Queue* q) {
    return q->size == 0;
}

// return size of queue
int size(Queue* q) {
    return q->size;
}

// peek at front element
int peek(Queue* q) {
    if (isEmpty(q)) {
        printf("Queue is empty!\n");
        return -1;
    }
    return q->arr[q->front];
}

// print queue contents
void printQueue(Queue* q) {
    if (isEmpty(q)) {
        printf("Queue is empty!\n");
        return;
    }
    int count = q->size;
    int i = q->front;
    while (count > 0) {
        printf("%d -> ", q->arr[i]);
        i = (i + 1) % q->capacity;
        count--;
    }
    printf("null\n");
}

// free memory
void freeQueue(Queue* q) {
    if (q) {
        free(q->arr);
        free(q);
    }
}

int main() {
    Queue* q1 = createQueue(10);

    enqueue(q1, 6);
    enqueue(q1, 1);
    enqueue(q1, 9);
    enqueue(q1, 2);
    enqueue(q1, 5);

    printQueue(q1);
    printf("Size: %d\n", size(q1));
    printf("Dequeue: %d\n", dequeue(q1));
    printf("isEmpty: %d\n", isEmpty(q1));

    printQueue(q1);
    printf("Size: %d\n", size(q1));
    printf("Peek: %d\n", peek(q1));

    freeQueue(q1);
    return 0;
}